### PR TITLE
Fix Mongo topology resource release

### PR DESCRIPTION
**Pull request by smallinsky**
## Issue
https://github.com/gravitational/teleport/issues/10663

## What 
Add missing topology `Disconnect` call in order to release resources.

### CODE REVIEW

File: lib/srv/db/mongodb/engine.go [hunk 1]
```diff
@@ -80,11 +80,12 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
...
-	serverConn, err := e.connect(ctx, sessionCtx)
+	serverConn, closeFn, err := e.connect(ctx, sessionCtx)
 	if err != nil {
 		return trace.Wrap(err, "error connecting to the database")
 	}
 	defer func() {
+		closeFn()
 		err := serverConn.Close()
 		if err != nil {
 			e.Log.WithError(err).Error("Failed to close server connection.")
```
Comment (greedy52) [inline lib/srv/db/mongodb/engine.go:91]: nit: maybe we can consolidate this inside `closeFn`?

File: integration/db_integration_test.go [hunk 2]
```diff
@@ -210,6 +210,45 @@ func TestDatabaseAccessMongoRootCluster(t *testing.T) {
...
+	}
+	connectMongoClient(t)
+
+	// Get connection count after mongo driver indicated the connection pool.
+	initialConnectionCount := pack.root.mongo.GetActiveConnectionsCount()
+	clientCount := 8
+	for i := 0; i < clientCount; i++ {
+		connectMongoClient(t)
+	}
+	// Check if active connections count is not growing over time when  new clients connects to the mongo server.
```
Comment (greedy52) [inline integration/db_integration_test.go:248]:
```suggestion
	// Check if active connections count is not growing over time when new clients connect to the mongo server.
```

**Review by greedy52 — APPROVED** [review]

File: lib/srv/db/mongodb/connect.go [hunk 3]
```diff
@@ -41,33 +41,39 @@ import (
...
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, nil, trace.Wrap(err)
+	}
+
+	closeFn := func() {
+		if err := top.Disconnect(ctx); err != nil {
+			e.Log.Warn("Failed to close topology: %v", err)
+		}
 	}
```
Comment (r0mant) [inline lib/srv/db/mongodb/connect.go:75]: Why not make the closeFn to close both `conn` and topology so the caller just needs to call it and doesn't have to call conn.Close separately?

File: lib/srv/db/mongodb/engine.go [hunk 4]
```diff
@@ -80,11 +80,12 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
...
 	}
 	// Establish connection to the MongoDB server.
-	serverConn, err := e.connect(ctx, sessionCtx)
+	serverConn, closeFn, err := e.connect(ctx, sessionCtx)
 	if err != nil {
 		return trace.Wrap(err, "error connecting to the database")
 	}
 	defer func() {
+		closeFn()
 		err := serverConn.Close()
```
Comment (r0mant) [inline lib/srv/db/mongodb/engine.go:89]: As mentioned above, I'd roll this into closeFn too.

File: lib/srv/db/mongodb/test.go [hunk 5]
```diff
@@ -257,6 +261,10 @@ func (s *TestServer) Port() string {
 	return s.port
 }
 
+func (s *TestServer) GetActiveConnectionsCount() int32 {
```
Comment (r0mant) [inline lib/srv/db/mongodb/test.go:264]: Nit: Add godoc.

**Review by r0mant — APPROVED** [review]
lgtm with a comment

**Review by greedy52 — APPROVED** [review]

**Comment by russjones** [comment]
@smallinsky Looks like this is approved, can you merge?

**Pull request merged**